Simplify value

joneshf-cn opened the pull request:
##### What it does

Through discussions, we realized that the value had more cases than we wanted to handle. We changed the types here to reflect what we want to handle.

Also sets up some round-trip tests for the JSON decoding/encoding. Turns out we weren't doing any decoding of prints.

joneshf-cn committed:
Remove some cases from values

Looks like we had 14 different cases to handle for rendering.
Most of these weren't cases that existed in our mental model.

Changing the types around a bit,
we were able to drop the cases that didn't make sense.

This should also make things a little easier for users to create.

joneshf-cn committed:
Add a test for `InputSource _` decoding/encoding

Since we're changing this stuff,
we probably want to make sure we don't introduce weird behaviors.
In this case,
we want to validate that encoding then decoding works the way we expect.

joneshf-cn committed:
Add `InputSource _` to the JSON

We've changed `InputSource _` and how it translates to JSON.
To avoid drift, we add it to the JSON tests.

joneshf-cn committed:
Test more round tripping

The test for `Input Expr` actually fleshed out an issue:
we weren't roundtripping `Expr` properly.

We added a test for that and were able to get that fixed.

davezuch commented on src/Data/Form.purs:
```diff
@@ -35,7 +39,7 @@ type Field f = Record (FieldRows f ())
 type SharedRows f r =
   ( default :: Maybe f
-  , value :: Record (InputState f ())
+  , value :: Maybe (InputSource f)
```
I was thinking a little more about how we're modeling this. Ultimately we have three possible value here
```purescript
Nothing | Just (UserInput a) | Just (Invalid a)
```
I'm just slightly bothered by having an invalid state in right-hand side (`Just` case). Seems like we could get the same results by rearranging it like
```purescript
Left NotSet | Left (Invalid a) | Right a
```
The benefit here is better separation between valid and invalid values. We can use `map` to only map over valid values, or `isLeft` to know if our value is invalid and we should resort to our `default`.

I wonder what adding validation support will do here, if anything. Not sure yet if it should be coupled to this value or not, but I can imagine expanding the sum type we use on the left for other possible error states.

joneshf-cn replied:
I'm with it. Maybe instead of bringing in the connotations of `Maybe _`/`Either _ _`, we enumerate the cases at our domain level?
```PureScript
data Value a
  = NotSet
  | Invalid a
  | UserInput a
```

My reasoning is, `NotSet` should only be an error if the field is required and there's no default value. If the field isn't required or there's a default value, it seems fine to have the value be `NotSet`.

What do you think?

davezuch replied:
Hmm, yeah you're right `NotSet` isn't necessarily an error, but there's still a distinction between having a value and not. If we drop the `Either` we lose the bifunctor machinery, but maybe that's okay. We can always add it back if we decide it'd be more useful or a better representation, refactoring is cheap 😉

joneshf-cn replied:
K. I'll switch it over and merge afterward.

joneshf-cn committed:
Merge remote-tracking branch 'origin/master' into HEAD

joneshf-cn committed:
Drop the `Maybe _` around `InputSource _`

We were using `Maybe (InputSource _)` to represent three things:
* No value is set
* The value is set by the user
* The value is invalid
There were two conflicting ideas here.

One is that if the value is set (whether valid or invalid),
we were lumping them into the "successful" part of the `Maybe _`.
If we're looking to tell whether a required field will validate,
we have to check both whether it exists and whether it's not invalid.

The other is that if the field is not required and we want to tell
whether it will validate,
we have to check either that the value doesn't exist or the value
is not invalid.

Using something like:
```PureScript
data ValueError a
  = NotSet
  | Invalid a

type Value a = Either (ValueError a) a
```
would address the first case, but not the second.

Another shifting of `NotSet` to the "successful" part of `Either _ _`
would address the second case, but not the first.

Rather than bring in the connotations of `Maybe _` or `Either _ _`,
we give it a name relevant to our domain.
Hopefully, this should be clearer to everyone involed.

It does mean that we'll have to write explicit functions for validation,
but that doesn't seem like the worst thing (given our options).

davezuch reviewed (approved):
Solid!
